File: BraintreeCore/src/main/java/com/braintreepayments/api/PreferredPaymentMethodsResult.java
package com.braintreepayments.api;

import org.json.JSONException;
import org.json.JSONObject;

/**
 * Contains information about which payment methods are preferred on the device.
 * This class is currently in beta and may change in future releases.
 */
public class PreferredPaymentMethodsResult {
    private boolean mPayPalPreferred;
    private boolean mVenmoPreferred;

    PreferredPaymentMethodsResult() {
    }

    static PreferredPaymentMethodsResult fromJSON(String responseBody, boolean venmoInstalled) {
        boolean payPalPreferred = false;

        try {
            JSONObject response = new JSONObject(responseBody);
            JSONObject preferredPaymentMethods = getObjectAtKeyPath(response, "data.preferredPaymentMethods");
            if (preferredPaymentMethods != null) {
                payPalPreferred = preferredPaymentMethods.getBoolean("paypalPreferred");
            }
        } catch (JSONException ignored) {
            // do nothing
        }

        PreferredPaymentMethodsResult result = new PreferredPaymentMethodsResult();
        result.isPayPalPreferred(payPalPreferred);
        result.isVenmoPreferred(venmoInstalled);
        return result;
    }

    public void isPayPalPreferred(boolean payPalPreferred) {
        mPayPalPreferred = payPalPreferred;
    }

    public void isVenmoPreferred(boolean venmoPreferred) {
        mVenmoPreferred = venmoPreferred;
    }

    /**
     * @return True if PayPal is a preferred payment method. False otherwise.
     */
    public boolean isPayPalPreferred() {
        return mPayPalPreferred;
    }

    /**
     * @return True if Venmo app is installed. False otherwise.
     */
    public boolean isVenmoPreferred() {
        return mVenmoPreferred;
    }

    private static JSONObject getObjectAtKeyPath(JSONObject obj, String keyPath) throws JSONException {
        String[] keys = keyPath.split("\\.");
        JSONObject result = obj;
        for (String key : keys) {
            result = result.getJSONObject(key);
        }
        return result;
    }
}

File: BraintreeCore/src/main/java/com/braintreepayments/api/PreferredPaymentMethodsResult.kt
package com.braintreepayments.api

import androidx.annotation.RestrictTo
import org.json.JSONException
import org.json.JSONObject

//NEXT MAJOR VERSION: constructor with optional params
/**
 * Contains information about which payment methods are preferred on the device.
 * This class is currently in beta and may be removed in future releases.
 * @hide
 */
open class PreferredPaymentMethodsResult @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP) constructor() {

    private var isPayPalPreferred = false
    private var isVenmoPreferred = false

    /**
     *
     * @return True if PayPal is a preferred payment method. False otherwise.
     */
    open fun isPayPalPreferred() = isPayPalPreferred

    /**
     *
     * @return True if Venmo app is installed. False otherwise.
     */
    open fun isVenmoPreferred() = isVenmoPreferred

    open fun isPayPalPreferred(payPalPreferred: Boolean): PreferredPaymentMethodsResult {
        isPayPalPreferred = payPalPreferred
        return this
    }

    open fun isVenmoPreferred(venmoPreferred: Boolean): PreferredPaymentMethodsResult {
        isVenmoPreferred = venmoPreferred
        return this
    }

    companion object {
        @JvmStatic
        fun fromJSON(
            responseBody: String,
            venmoInstalled: Boolean
        ): PreferredPaymentMethodsResult {
            var payPalPreferred = false
            try {
                val response = JSONObject(responseBody)
                val preferredPaymentMethods =
                    getObjectAtKeyPath(response, "data.preferredPaymentMethods")
                if (preferredPaymentMethods != null) { //NEXT MAJOR VERSION: this shouldn't be null
                    payPalPreferred = preferredPaymentMethods.getBoolean("paypalPreferred")
                }
            } catch (ignored: JSONException) {
                // do nothing
                print(ignored.message)
            }
            return PreferredPaymentMethodsResult()
                .isPayPalPreferred(payPalPreferred)
                .isVenmoPreferred(venmoInstalled)
        }

        @Throws(JSONException::class)
        private fun getObjectAtKeyPath(obj: JSONObject, keyPath: String): JSONObject {
            val keys = keyPath.split(".").toTypedArray()
            var result = obj
            for (key in keys) {
                result = result.getJSONObject(key)
            }
            return result
        }
    }
}